### ISSUE TITLE
New default theme 'Bootstrap 4', Part 2

### ISSUE THREAD

**Issue by juek** [Typesetter/Typesetter#672]
Continuing [this issue](https://github.com/Typesetter/Typesetter/issues/664) in a new thread

**Comment by juek** [comment]
> It would be great to have an export/import to save work before experiments and reload after fail.

Any ideas for the UI for this?

**Comment by mahotilo** [comment]
> Any ideas for the UI for that?

Minimal UI like for file attachment to email. 3 Button: select local file name, export, import

**Comment by juek** [comment]
> Or could such way cause security problems?

Anything that can upload executable code (like config.php) is a no-go. But using JSON would be an option.

**Comment by gtbu** [comment]
I j ust had a look at the momentary master. The offcanvas nav is "ok" , but it is 100 % broad - an so the seek is too broad and also the rifgr side. In the [bootstrap 4.5 examples is an offcanvas](https://getbootstrap.com/docs/4.5/examples/) , at wich i can set .offcanvas-collapse.open to width: 50% to solve that. Here it is different (media width  and header class="main-header ).

**Comment by juek** [comment]
> The "offcanvas nav" is 100 % broad

Hbb, i haven't jet pushed anything from the alternative mobile menu styles, but it's in the works. So all we currently have in the master is the default **pulldown** nav.

> https://github.com/vmitsaras/js-offcanvas 

Yes, that's basically what will be selectable.

**Comment by juek** [comment]
> https://getbootstrap.com/docs/4.5/examples/offcanvas

This variant will be called 'Slide Over' in our theme. 'Off Canvas' will be similar but 'push' the rest of the page off the viewport.

**Comment by gtbu** [comment]
I see i was in too much of a hurry. In my (coming) BT5biz - Template the offcanvas destroys the sticky nav. There must be some incompatible code in the css(needs restyling and evtl some js-code) . Thats why i prefer the js-offcanvas, which needs jquery (or wait and  have a look at Your solution).
PS. The dropwdows have still a for my taste to big white border - though probably a bootstrap-style which can be changed in the custom.css ( https://behigh.github.io/bootstrap_dropdowns_enhancement/#submenu  - for what i mean)

**Comment by juek** [comment]
> In my (coming) BT5biz - Template the offcanvas destroys the sticky nav. There must be some incompatible code in the css(needs restyling and evtl some js-code) 

Sounds like you have to take care not to load any js or css assets unrelated to the current theme/layout. 

> Thats why i prefer the js-offcanvas, which needs jquery 

Typesetter will continue to use jQuery for an indefinite period – so it is there even with your Bootstrap 5 theme. 
You don't have to worry about.

**Comment by juek** [comment]
> The dropwdowns have still a for my taste a to too big white border

Can be adjusted in variables.scss or via the Scss tab in Layout Editor. Find and change…

```scss
$dropdown-padding-y: 0.5rem;
$dropdown-item-padding-y: 0.25rem;
$dropdown-item-padding-x: 1.5rem;
```

**Comment by juek** [comment]
> the offcanvas destroys the sticky nav. There must be some incompatible code in the css

FYI: as [noted earlier](https://github.com/Typesetter/Typesetter/issues/664#issuecomment-683278582) position:sticky does not work in overflow:hidden context. Most off-canvas solutions apply overflow-hidden to the body element. 
Another little known fact is that fixed positioning will not work inside transform context. We may transform fixed elements themselves but they may not be inside transformed elements. All off-canvas solutions I know out there use transform:translate on the html/body elements. We cannot do it like that.

That's the main reason why I'm still working on the different menu styles – quite tricky to get them to work with all possible settings. But I'm almost done.

**Comment by juek** [comment]
Here come the 'Hamburger Menu Styles' with https://github.com/Typesetter/Typesetter/commit/bba14f421844d69a8478260b501ced50d03f9524

First, there are a few other changes too…

* added font 'Source Serif Pro' to customizer
* added JS breakpoint detection which fires [custom events](https://github.com/Typesetter/Typesetter/commit/bba14f421844d69a8478260b501ced50d03f9524) 'breakpoint_change', 'navbar_expanded' and 'navbar_collapsed'. You can also check out the 'Theme_Bootstrap4' JS object in custom scripts. The &lt;html&gt; now gets additional CSS classes dynamically assigned via JS – look for stuff like 'breakpoint-xs' and 'navbar-expanded' when changing the viewport width.

For the mobile menu styles:

They certainly still need some tweaks. I haven't yet tested all relevant customizer combinations. I'll probably also add a thumb-friendly 'Slide Up' style, which is currently trending and makes sense. 

The animated hamburger is a bit playful and maybe not everyone's taste. Although there is [already a Scss variable](https://github.com/Typesetter/Typesetter/blob/master/themes/Bootstrap4/_common/mobile_menu.scss#L17e, I yet haven't putit into customizer. Shouold we add it? Tell me what you think.

**Comment by juek** [comment]
Important! 

I forgot to mention that 'Off Canvas' and 'Slide Over' menus can be swiped away, but that requires jquery-touch which is only available since [this commit](https://github.com/Typesetter/Typesetter/commit/5a6e9ec644199baa9d2e3b4938bf4076c7d83ac1)

You will see JS errors and non-working menus without it!

**Comment by mahotilo** [comment]
It seems only to work for *footer*. Can you check it, please.

**Comment by mahotilo** [comment]
Slide Over appears and hides on window resize, e.g. on screen rotation.

**Comment by juek** [comment]
> It seems only to work for footer. Could you check it, please.

Yep, fix is here https://github.com/Typesetter/Typesetter/commit/5b76151d799831073d1bfe034de90af1d29026d7

**Comment by juek** [comment]
> Slide Over/Off-Canvas appears and hides on window resize, e.g. on screen rotation.

Yep, that definitely still needs some tweaking.

**Comment by juek** [comment]
> Yep, that definitely still needs some tweaking.

I should trust the DONT-YOU-WRITE-INLINE-CSS bell ringing. What a whole lotta mess. This is JS event fireworks! m(

Seems like I [fixed it](https://github.com/Typesetter/Typesetter/commit/3a461bedb865b0dadfc98127386d21bcd643749d)

**Comment by gtbu** [comment]
Maybe i overlooked it : Is the position of the main menu fix attached to the searchfield   und not (yet) adjustable  in the middle ?  (like d-flex justify-content-between )

The searchfunction m,ay look a bit more stylish  as in https://codepen.io/gungorbudak/pen/ooKNpz (1st line)

The Login and sitemap may have an icon instead .

**Comment by juek** [comment]
> … main menu … (yet) adjustable in the middle?

No. At least not yet. Currently we need to change [this line in template.php](https://github.com/Typesetter/Typesetter/blob/master/themes/Bootstrap4/footer/template.php#L82)

> … searchfunction may look a bit more stylish … The Login and sitemap may have an icon instead

Lots of things could be more stylish, but it depends on what different people think is stylish. 
I'm just trying to stick as close to the bootstrap defaults as possible. At least that was the original goal.

**Comment by juek** [comment]
> … main menu … (yet) adjustable in the middle?
> No. At least not yet. Currently we need to change this line in template.php

It's really easy to implement - so yeah, yet's do it.

**Comment by juek** [comment]
> … main menu adjustable in the middle?

Done with https://github.com/Typesetter/Typesetter/commit/3021cc63d6398a35a09c1bebf8ef92f8c87aa2fb

**Comment by juek** [comment]
> The searchfunction may look a bit more stylish …

I would let myself be persuaded to do this:

![grafik](https://user-images.githubusercontent.com/704283/93006032-a789b080-f557-11ea-81c4-3246ca9ab6e0.png)

Thumbs please

**Comment by juek** [comment]
I ended up with this https://github.com/Typesetter/Typesetter/commit/d1be675f46647c7033cb0170c73bd3331e2a5b0e

![grafik](https://user-images.githubusercontent.com/704283/93006306-d2293880-f55a-11ea-842d-40fb84e3f9fd.png)

But we can still discuss everything.
Overall, the functional part of the theme is almost done and I want to start customizer tests to see how it works in practice. And how far we can actually get with it.

While I like the current dark header colors, fresher, friendlier colors might make better default settings. 
It really depends on the content and the CI / logo etc. in the end.

I would love to see some great examples from you guys!

**Comment by juek** [comment]
OMG

![grafik](https://user-images.githubusercontent.com/704283/93007800-51276c80-f56d-11ea-8826-3142dd184d1e.png)

Using `@extend` can add insane overhead. This is ~150 KB too much CSS.

**Comment by mahotilo** [comment]
> While I like the current dark header colors, fresher, friendlier colors might make better default settings.
> It really depends on the content and the CI / logo etc. in the end.

I like them too. And this black theme appears to perfectly fit one abandoned project about industrial rubber that I use as a playground:)
I can only suggest adding the theme's BS4 logo as the default Brand Logo value.

And we have a warning from Lighthous about admin links default colors and white background.
![image](https://user-images.githubusercontent.com/14929385/93011514-7c2ab400-f59f-11ea-86fc-cf1f5ae38fe0.png)

**Comment by mahotilo** [comment]
> I ended up with this d1be675
> But we can still discuss everything.

I dare to propose one more search setting "Size". One option - big, the other - compact. Because for mobile first sites, the BS default large fields fit better.

**Comment by gtbu** [comment]
Well - if You look for a big navbar - You can try  https://bootsnipp.com/snippets/a6aV0

**Comment by juek** [comment]
> And we have a warning from Lighthous about admin links default colors and white background.

I assume they simply take CIELab L50 (psychometric lightness) as a threshold – which is not all bad.

![grafik](https://user-images.githubusercontent.com/704283/93016105-01719100-f5bf-11ea-83c2-493833cc1166.png)

If so, they just barely passed their own test :)
However, If that's a ranking factor, we should actually make it that (unrecognizable) tad darker ;)

**Comment by mahotilo** [comment]
I have found a conflicting pair

![image](https://user-images.githubusercontent.com/14929385/93016308-ed329180-f5c8-11ea-90d1-d8c2d18f7bdf.png)

**Comment by juek** [comment]
> I have found a conflicting pair

Can't confirm. Scss compilation must have gone wrong. Please clear CMS chache and try again

**Comment by juek** [comment]
> I have found a conflicting pair

Fixed with https://github.com/Typesetter/Typesetter/commit/5ff7aa1ada68615486c8e835f52366acf905ebe1

**Comment by gtbu** [comment]
* After playing around i was not able to influence the hover color of the menu
![rc52hover](https://user-images.githubusercontent.com/30436307/93057474-456f9f00-f66e-11ea-9d5b-ae49ec2e15b2.jpg)

* Will You add a menu hover function *

**Comment by juek** [comment]
> Will You still add a menu hover function ?

Yes, I'm afraid that we also need dropdown colors customizable. 
IMO, Customizer slowly gets bloated and loses the advantages it has over the scss editor.

**Comment by gtbu** [comment]
I can understand the work  for You that is behind it , but i thinkcustomizer is quite comfortable for users. Only some bugs may need a quick fix - so never mind (...or take refuge...).

**Comment by mahotilo** [comment]
> IMO, Customizer slowly gets bloated and loses the advantages it has over the scss editor.

In my opinion, the power of this template is not in its full customization, but in a balanced list of options allowing beginners to get dozens of cool and almost unique designs. 
I guess, the current options list is nearing its logical end, and the goal is almost achieved.

**Comment by gtbu** [comment]
The SCSS - problem may be solved, but the problem that the submenu has a blue li while the top li is black(or another color) is still unsolved. Something like
.dropdown-item, .dropdown-item:active {background:yellow !important;}  helps here

**Comment by juek** [comment]
@gtbu take it easy, several things are not yet finished

**Comment by juek** [comment]
> OMG, using @extend can add insane overhead. This is ~150 KB too much CSS.

latest commits (Sep 18th) saved us some 120 KB

Before
![grafik](https://user-images.githubusercontent.com/704283/93605426-07081600-f9c7-11ea-99e2-f3c2f11524d8.png)

After
![grafik](https://user-images.githubusercontent.com/704283/93605395-fa83bd80-f9c6-11ea-89a0-f373b44c0545.png)

**Comment by juek** [comment]
> I can only suggest adding the theme's BS4 logo as the default Brad Logo value.

Well it turned out to be this https://github.com/Typesetter/Typesetter/commit/7927bd03a20c40a13d99caf1d5ba43b255848d50 https://github.com/Typesetter/Typesetter/commit/65297c6b659eb40feee8870123b44af0af8d8601

![grafik](https://user-images.githubusercontent.com/704283/93615481-bd71f800-f9d3-11ea-973a-be48af0579df.png)

![Typesetter_Logo_2020](https://user-images.githubusercontent.com/704283/93615487-c1057f00-f9d3-11ea-88ed-0b7c2e62e16e.png)

**Comment by mahotilo** [comment]
A new "t"? Rebranding or just a theme logo?

**Comment by mahotilo** [comment]
OMG
![image](https://user-images.githubusercontent.com/14929385/93631937-20c05200-f9f5-11ea-97b4-2532f4e054ec.png)

**Comment by juek** [comment]
> It seems too thin to me and my display

It's meant to be thin or better call it light.

> Rebranding or just a theme logo?

Well, we do not actually have a real logo or 'trademark'. Nobody would pay the protection fees. 
The current [square black logotype](https://avatars2.githubusercontent.com/u/16497604?s=200&v=4) looks too traditional to me, and, well, too black. So here is a new one. A (t) in a circle. Nothing we haven't already seen a thousand times. 
Bootstrap is a [B] in a square, pretty similar stuff. Nothing to build a philosophical way of thinking about. And hard to argue about.

**Comment by juek** [comment]
> It is in SVG and CSS

In fact, not yet tested.

**Comment by mahotilo** [comment]
> It's meant to be thin or better call it 'light'.

I am not talking about design itself but about output. I just want to say, that this svg logo looks stepped and not smooth. Maybe it is my Opera or my display or graphics. But a thicker line usually helps me :)

**Comment by juek** [comment]
> Maybe some CSS issues?

Looks OK here on Opera

![grafik](https://user-images.githubusercontent.com/704283/93641964-8c0e2200-f9fd-11ea-9e9e-65cd04bf1aa3.png)

Default size fixed with https://github.com/Typesetter/Typesetter/commit/158f4bb421ede6650572466e08ed51f54e358d6a

**Comment by mahotilo** [comment]
> Looks OK here on Opera

I have 125% scale in Win7. Maybe it somehow affects svg rendering. I don't know. Anyway, with the last fix, the logo looks better to me.

**Comment by juek** [comment]
> Will You still add a menu hover function ?

Dropdown menu customization added with https://github.com/Typesetter/Typesetter/commit/9348998c4a61293dae51ff3cc76b60540497f2b3

With that amount of color settings, it should be possible to define – let's say 6 – basic colors and show them as swatches next to every color picker field. 
But, maybe in an update. I've already had enough!

**Comment by mahotilo** [comment]
We have some conflict with modal dialog.
![image](https://user-images.githubusercontent.com/14929385/94358948-e9187080-00ac-11eb-81ed-d259c24865fa.png)

What was the purpose for this?
https://github.com/Typesetter/Typesetter/blob/11f9882f8882212909df1f2697be331d94ce5cd4/themes/Bootstrap4/_common/customizer_common_headers.scss#L194

**Comment by mahotilo** [comment]
One more note. There is no padding in the drop-down list of classes.

![image](https://user-images.githubusercontent.com/14929385/94366920-d2d9d700-00e3-11eb-907f-25d8fdeebc16.png)

**Comment by juek** [comment]
> What was the purpose for this?

That's so that the content doesn't scroll over the headers. [I changed it to use the Bootstrap variables](https://github.com/Typesetter/Typesetter/commit/a6fddeffdcaace190baa41abb8854903a4b03fd6). That's what we have it for ;)

**Comment by juek** [comment]
> There is no padding in the drop-down list of classes.

Never was. The actual dropdown only starts next to the checkbox.

**Comment by mahotilo** [comment]
> Never was. 

Yep. My bad

**Comment by mahotilo** [comment]
> That's so that the content doesn't scroll over the headers. I changed it to use the Bootstrap variables. 

Yes, this helps.

**Comment by mahotilo** [comment]
I can't catch the moment, but sometimes some changes in layout cause the complementary header to change its state from show: never to show: always. Have you ever faced this issue?

**Comment by Sestowner** [comment]
> ...sometimes some changes in layout cause the complementary header to change its state from show: never to show: always.

Should be fixed with this PR #682.
